R/path_linters.R

```diff
@@ -0,0 +1,208 @@
+
+is_absolute_path <- function(str) {
+  re_matches(str, absolute_path_regex)
+}
+
+is_root_path <- function(str) {
+  re_matches(str, root_path_regex)
+}
+
+is_relative_path <- function(str) {
+  re_matches(str, relative_path_regex)
+}
+
+is_path <- function(str) {
+  re_matches(str, path_regex)
+}
+
+is_valid_path <- function(str, lax=FALSE) {
+  # Given a character vector of paths, return FALSE for directory or file having valid characters.
+  # On Windows, invalid chars are all control chars and: * ? " < > | :
+  # On Unix, all characters are valid, except when lax=TRUE (use same invalid chars as Windows).
+  as.logical(
+    Map(
+      function(dirs, is_win32) {
+        if (is_win32 || lax) {
+          if (length(dirs) && is_root_path(dirs[[1L]])) {
+            dirs <- tail(dirs, -1L)  # remove root element ("/", "C:", or "\\")
+          }
+          !any(re_matches(dirs, unsafe_char_regex))
+        } else {
+          TRUE  # either Unix path or strict (lax=FALSE)
+        }
+      },
+      split_path(str),
+      re_matches(str, rex(or(list(root_regex[["win32"]], root_regex[["unc"]], "\\"))))
+    )
+  )
+}
+
+is_long_path <- function(str) {
+  # TRUE , e.g.: "./foo", "C:\\foo", "foo/bar"
+  # FALSE, e.g.: "/",  "\\", "n/a", "/foo", "foo/"
+  re_matches(
+    re_substitutes(str, ":", ""),
+    rex(at_least(safe_char_regex, 1L), one_of("/", "\\"), at_least(safe_char_regex, 2L))
+  )
+}
+
+is_valid_long_path <- function(str, lax=FALSE) {
+  # Convenience function to avoid linting short paths and those unlikely to be valid paths
+  ret <- is_valid_path(str, lax)
+  if (lax) {
+    ret <- ret & is_long_path(str)
+  }
+  ret
+}
+
+
+split_path <- function(fp, fsep="/|\\\\") {
+  if (!is.character(fp)) {
+    stop("argument 'fp' should be a character vector")
+  }
+  if (!is.character(fsep) || length(fsep) != 1L || !nzchar(fsep)) {
+    stop("argument 'fsep' should be a non-empty regular expression character string")
+  }
+  Map(
+    function(dirs, prefix) {
+      # add root dir if needed
+      i <- 1L
+      for (dir in seq_along(dirs)) {
+        if (!nzchar(dirs[[i]])) {
+          i <- i + 1L
+        } else {
+          break
+        }
+      }
+      i <- i - 1L
+      if (i > 0L) {
+        dirs <- c(paste0(rep(prefix, i), collapse=""), tail(dirs, -i))
+      }
+      # add // to protocols (like http, smb, ...)
+      if (length(dirs)) {
+        l <- nchar(dirs[[1L]])
+        if (l > 2L && substr(dirs[[1L]], l, l) == ":") {
+          dirs[[1L]] <- paste0(dirs[[1L]], "//")
+        }
+      }
+      # remove empty dirs
+      dirs[nzchar(dirs)]
+    },
+    strsplit(fp, fsep),
+    substr(fp, 1L, 1L)
+  )
+}
+
+
+unquote <- function(str, type=c("'", "\"", "`")) {
+  # Remove the outtermost quotes and unescape backslashes
+  for (q in type) {
+    str <- re_substitutes(str, rex(start, q, capture(zero_or_more(any)), q, end), "\\1")
+  }
+  stringi::stri_unescape_unicode(str)
```

Comment:
We cannot depend on stringi, you can look at base alternatives if you need this http://stackoverflow.com/a/25310392/2055486